File: String/3598. Longest Common Prefix Between Adjacent Strings After Removals.py
class Solution:
    def longestCommonPrefix(self, words: List[str]) -> List[int]:

        n = len(words)
        suffix = [0] * n
        prefix = [0] * n

        if n == 1:
            return [0]

        @lru_cache(None)    
        def lcp(s1, s2):
         
            res = 0
            for a, b in zip(s1, s2):
                if a != b: return res
                res+=1
            return res

            
        for i in range(n-2, -1, -1):
            suffix[i] = max(suffix[i+1], lcp(words[i], words[i+1]))

        for i in range(1, n):
            prefix[i] = max(prefix[i-1], lcp(words[i], words[i-1]))

        ans = []
        
        for i in range(n):
            if i-1 < 0:
                ans.append(suffix[i+1])
                continue
                
            if i+1 == n:
                ans.append(prefix[i-1])
                continue

            ans.append(max(prefix[i-1], suffix[i+1], lcp(words[i-1], words[i+1])))

        return ans

class Solution:
    def longestCommonPrefix(self, words: List[str]) -> List[int]:
        def lcp(i, j):
            for k, (w_a, w_b) in enumerate(zip(words[i], words[j])):
                if w_a !=  w_b:
                    return k
            return min(len(words[i]), len(words[j]))
        
        L = [lcp(i, i+1) for i in range(len(words)-1)]
        cur_lcp = list(accumulate(L, max, initial=0))
        rev_lcp = list(accumulate(L[::-1], max, initial=0))[::-1]

        res = [0] * len(words)
        if len(words) == 1 :
            return res
        res[0] = rev_lcp[1]
        res[-1] = cur_lcp[-2]
        for i in range(1,len(words)-1):
            res[i] = max(cur_lcp[i-1], rev_lcp[i+1], lcp(i-1, i+1))
        return res

File: String/3598. Longest Common Prefix Between Adjacent Strings After Removals.cpp
/**
 * @brief 
 * 
 *  prefix[i] store the max between combination of i-1 and i or prefix[i-1]
 *  suffix[i] store the max between combination of i and i+1 or suffix[i+1]
 * 
 */
class Solution {
public:
    vector<int> longestCommonPrefix(vector<string>& words) {
        function<int(int, int)> lcp = [&](int i, int j){
            for(int k = 0; k < min(words[i].length(), words[j].length()); ++k){
                if (words[i][k] != words[j][k])
                    return k;
            }
            return min((int)words[i].length(), (int)words[j].length());
        };

        int n = words.size();
        vector<int>res(n); 
        if (n == 1) return res;
        vector<int>suffix(n);
        vector<int>prefix(n);
        for(int i = 1; i < n - 1; ++i){
            prefix[i] = max(prefix[i-1], lcp(i, i-1));
        }
        for(int i = n - 2; i>0; --i){
            suffix[i] = max(suffix[i+1], lcp(i, i+1));
        }

        res[0] = suffix[1];
        res[n-1] = prefix[n-2];
        for(int i = 1; i<n-1; ++i){
            res[i] = max(prefix[i-1], max(suffix[i+1], lcp(i-1, i+1)));
        }
        return res;
    }
};